### PR TITLE
gmm_v2: add transpose_rhs so callers can skip the swapaxes on the weight

### PR DESCRIPTION
**Pull request by littlehau**
`dlhs = dy @ W^T` contracts `W`'s `n` axis. `W` is stored `[group, k, n]`, and the only way to express that before this change was `rhs.swapaxes(1, 2)` — free in JAX, but not at a Pallas boundary: **Pallas pins its operands' layout**, so XLA cannot fuse a transpose into a kernel operand and materializes a second copy of the whole weight. `transpose_rhs=True` instead flips the BlockSpec index map to `(group, n_id, k_id)` and reads the existing buffer. Zero extra bytes.

## Ships with its first consumer

`ragged_dot`'s own dlhs path was doing exactly that `swapaxes`, so this deletes it:

```diff
 elif ragged_dot_dimension_numbers == DLHS_RAGGED_DOT_DIM_NUMS:  # dlhs
     out = gmm_backend.gmm_v2(
         lhs,  # [m, n]
-        rhs.swapaxes(1, 2),  # [num_groups, n, k]
+        rhs,  # [num_groups, k, n], addressed in place
         ...
+        transpose_rhs=True,
     )
```

Measured through `PallasMosaicTpuV2RaggedDot` on v7x, jax 0.11, bf16, `m=65536 g=128 k=3072 n=2048`, median of 20, µs/call:

| router | `swapaxes` | `transpose_rhs` | speedup |
|---|---:|---:|---:|
| balanced | 2214.2 | **1065.6** | 2.08× |
| skew50 | 2466.6 | **1317.5** | 1.87× |
| padworst | 2428.1 | **1304.7** | 1.86× |

Within **0.4%** of calling `gmm_v2` directly with the flag, so the Op layer adds no measurable overhead.

## It is a package with a large enough `tile_m`

The transpose is relocated, not removed — the MXU still wants the contraction on its preferred axis, so it happens on-chip per sub-tile at a cost scaling as `1/tile_m`. At `tile_m=128` the flag is a **net regression** (4553 vs 3399 µs); it only pays once `tile_m` reaches the MXU row granularity. `calculate_tiling` returns 768 for this shape, well clear of that.

## Safety

A static bool, so no runtime cost, but six sites must agree: `validate_inputs`' expected shape, `rhs_weight_index_map`, the BlockSpec, the `rhs_tile_n` axis, the tail-k mask's `k_axis`, and the matmul's `dimension_numbers`.

For a **direct caller** the flag is a promise, not a detection: `[G,k,n]` and `[G,n,k]` are indistinguishable by inspection, so a wrong claim at `k == n` silently returns wrong numbers. `rhs_scale`, `rhs_bias`, `lhs_scale`, `fuse_act` and a sub-8-bit `rhs` each raise `NotImplementedError` rather than being left to chance.

The **in-tree consumer is not exposed to that**: `DLHS_RAGGED_DOT_DIM_NUMS` pins `dot_dimension_numbers=(([1], [2]), …)`, i.e. contract rhs dim 2 — so the layout is determined by the dimension numbers rather than claimed by a caller.

Bit-exact: transposed and untransposed agree to `max|d| = 0` at `tile_m=512`.

### CODE REVIEW

**Comment by littlehau** [comment]
Closing this — c8df7c0 landed the same change (same flag name, same restrictions, and the identical
two-line switch in `pallas_mosaic_tpu_v2.py`'s dlhs path), and it ships with a test, which this PR did
not.

Leaving our measurements here in case they are useful, plus two observations on the landed version.

### The win, measured

Through `PallasMosaicTpuV2RaggedDot`'s dlhs path on TPU v7x, jax 0.11, bf16,
`m=65536 g=128 k=3072 n=2048`, µs/call, median of 20. Three synthetic `group_sizes`: `balanced` =
every group exactly `m/g`, `skew50` = `mean × U(0.5, 1.5)`, `padworst` = alternating `mean±1`.

| `group_sizes` | `swapaxes` | `transpose_rhs` | speedup |
|---|---:|---:|---:|
| balanced | 2214.2 | 1065.6 | 2.08× |
| skew50 | 2466.6 | 1317.5 | 1.87× |
| padworst | 2428.1 | 1304.7 | 1.86× |

Also 6/6 wins in fp8 (e5m2 cotangent × e4m3 weight), 2.15–2.41×, at the tiling `calculate_tiling`
selects for those shapes.

### 1. The flag's sign depends on `tile_m`, and it is now enabled unconditionally

`c8df7c0` sets `transpose_rhs=True` on every dlhs call, and does not touch `calculate_tiling`. But the
transpose is relocated rather than removed — it happens on-chip per sub-tile, so its cost amortizes as
`1/tile_m`. Forcing `tile_m=128` (the value `calculate_tiling` seeds, and keeps whenever the
bucket-growth loop cannot grow it — e.g. small `size_m`), same shapes as above:

| `group_sizes` | `swapaxes` | `transpose_rhs` | |
|---|---:|---:|---|
| balanced | 3402.2 | 4555.6 | 1.34× **slower** |
| skew50 | 3679.7 | 5141.6 | 1.40× **slower** |
| padworst | 3648.0 | 5084.7 | 1.39× **slower** |

At the tiling the heuristic actually picks for these shapes (`tile_m=768, bucket_base=256`) it is
1.40–1.47× *faster*, so the change is clearly right in the common case — but a caller whose `tile_m`
lands at 128 now silently regresses ~35–40% on dlhs. A `tile_m` threshold on the flag, or picking it
in `calculate_tiling` rather than at the call site, would close that.

### 2. `test_gmm_transpose_rhs` uses `in_size == out_size`, where the transpose is undetectable

The test parameterizes `in_size=512, out_size=512`, so `rhs` is `(num_groups, 512, 512)`. With `k == n`
the transposed and untransposed layouts have identical shapes, and `rhs_ref = jnp.swapaxes(rhs, 1, 2)`
is shape-compatible either way — so an incorrect index map or a wrong `dimension_numbers` still
produces a conformable matmul and can pass. `in_size != out_size` (e.g. 512 × 1024) would make that
class of error a shape error instead. Adding a non-zero `group_offset` case would also help; the
current product only covers `group_offset=0`.

### 3. Minor: the inner matmul could avoid the explicit `.T`

`c8df7c0` computes the transposed block as `jnp.matmul(lhs_blk, rhs_blk.T)`. Contracting on dim 1 of
both operands expresses the same thing without emitting a transpose:

```python
lax.dot_general(lhs_blk, rhs_blk, dimension_numbers=(((1,), (1,)), ((), ())),
                preferred_element_type=jnp.float32)
```

Whether Mosaic already folds the `.T` into the dot we have not measured, so this may be a no-op — but
if it does not, it is plausibly part of what makes the small-`tile_m` case above so expensive. Happy to
send a patch and numbers if that is of interest.
